src/actions/actionHeapSort.js

```diff
@@ -0,0 +1,190 @@
+  return {
+    type: 'START_HEAP_SORT',
+  }
+}
+
+export function heapSort(sortArray, heapSortArray, step2HeapSortArray, step3HeapSortArray, iterations, heapTreeTop, heapTreeRow2, heapTreeRow3, heapTreeRow4, heapSortStep, smallHeapSortStep, heapArrayIndex, startHeapTree, finishBuildHeap, isSorted = false) {
+  iterations += 1;
+  var isRunning = true;
+  var heapFinalSort = false;
+
+  // build the heap tree from orginal ramdom array
+  if ( startHeapTree === true ) {
+    var veiwedNumber = 0;
+    veiwedNumber = heapSortArray[heapSortStep];
+      if ( smallHeapSortStep === 1 ) {
+      heapTreeTop[0] = veiwedNumber;
+      }
+      if ( smallHeapSortStep === 2 ) {
+        if( heapSortStep === 0 ) heapTreeTop[5] = veiwedNumber;
+        if( heapSortStep === 1 ) heapTreeRow2[2] = veiwedNumber;
+        if( heapSortStep === 2 ) heapTreeRow2[8] = veiwedNumber;
+        if( heapSortStep === 3 ) heapTreeRow3[1] = veiwedNumber;
+        if( heapSortStep === 4 ) heapTreeRow3[3] = veiwedNumber;
+        if( heapSortStep === 5 ) heapTreeRow3[7] = veiwedNumber;
+        if( heapSortStep === 6 ) heapTreeRow3[9] = veiwedNumber;
+        if( heapSortStep === 7 ) heapTreeRow4[0] = veiwedNumber;
+        if( heapSortStep === 8 ) heapTreeRow4[1] = veiwedNumber;
+        if( heapSortStep === 9 ) heapTreeRow4[3] = veiwedNumber;
+      }
+      if ( smallHeapSortStep === 3 ) {
+        heapTreeTop[0] = -1;
+        heapSortArray[heapSortStep] = -1;
+      }
+      smallHeapSortStep += 1;
+      if ( smallHeapSortStep === 4 ) {
+        smallHeapSortStep = 1;
+        heapSortStep += 1;
+        if (heapSortStep > 9) {
+           startHeapTree = false;
+           heapSortStep = 0;
+           var step2Array = [ heapTreeTop[5], heapTreeRow2[2], heapTreeRow2[8], heapTreeRow3[1], heapTreeRow3[3], heapTreeRow3[7], heapTreeRow3[9], heapTreeRow4[0], heapTreeRow4[1], heapTreeRow4[3] ]
+           step2HeapSortArray = step2HeapSort(step2Array);
+           heapSortArray = step2HeapSortArray;
+        }
+      }
+  }
+  // sort the heap tree itself
+  if ( startHeapTree === false && finishBuildHeap === false ) {
+    veiwedNumber = 0;
+    veiwedNumber = step2HeapSortArray[heapSortStep];
+      if ( smallHeapSortStep === 1 ) {
+      heapTreeTop[9] = veiwedNumber;
+      }
+      if ( smallHeapSortStep === 2 ) {
+        if( heapSortStep === 0 ) heapTreeTop[5] = veiwedNumber;
+        if( heapSortStep === 1 ) heapTreeRow2[2] = veiwedNumber;
+        if( heapSortStep === 2 ) heapTreeRow2[8] = veiwedNumber;
+        if( heapSortStep === 3 ) heapTreeRow3[1] = veiwedNumber;
+        if( heapSortStep === 4 ) heapTreeRow3[3] = veiwedNumber;
+        if( heapSortStep === 5 ) heapTreeRow3[7] = veiwedNumber;
+        if( heapSortStep === 6 ) heapTreeRow3[9] = veiwedNumber;
+        if( heapSortStep === 7 ) heapTreeRow4[0] = veiwedNumber;
+        if( heapSortStep === 8 ) heapTreeRow4[1] = veiwedNumber;
+        if( heapSortStep === 9 ) heapTreeRow4[3] = veiwedNumber;
+      }
```

Comment:
We discussed that you might be able to DRY out this code more. You might want to at least add an issue to the GitHub issues to mark that it should be cleaned up later.